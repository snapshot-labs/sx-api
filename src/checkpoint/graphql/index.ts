import { RequestParamHandler } from 'express';
import { graphqlHTTP } from 'express-graphql';
import { GraphQLObjectType, GraphQLSchema } from 'graphql';

/**
 * Creates an graphql http handler for the query passed a parameters.
 * Returned middleware can be used with express.
 */
export default function get(query: GraphQLObjectType): RequestParamHandler {
  const schema = new GraphQLSchema({ query });
  return graphqlHTTP({ schema, graphiql: {} });
}
