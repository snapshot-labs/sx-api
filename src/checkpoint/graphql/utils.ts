import {
  introspectionFromSchema,
  IntrospectionNamedTypeRef,
  IntrospectionObjectType
} from 'graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { IResolvers } from '@graphql-tools/utils';
import mysql from '../mysql';

async function queryMulti(parent, args, context, info) {
  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(skip, first);
  return await mysql.queryAsync(
    `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ${orderBy} ${orderDirection} LIMIT ?, ?`,
    params
  );
}

async function querySingle(parent, args, context, info) {
  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

/**
 * Returns a list of objects defined within the graphql typedefs.
 * The types returns are introspection objects, that can be used
 * for inspecting the fields and types.
 *
 * Note, that the returned objects does not include the Query object type if defined.
 *
 */
const getObjectsSchemas = (typeDefs: string): IntrospectionObjectType[] => {
  const schema = makeExecutableSchema({ typeDefs: `type Query { x: String }\n${typeDefs}` });
  return introspectionFromSchema(schema)
    .__schema.types.filter(type => {
      return (
        type.kind === 'OBJECT' &&
        type.fields &&
        !type.name.startsWith('__') &&
        type.name !== 'Query'
      );
    })
    .map(type => type as IntrospectionObjectType);
};

/**
 * toSql generates SQL statements to create tables based on the types defined
 * in the GraphQL typeDefs.
 *
 * The generated SQL statment also creates a `checkpoint` table to track block
 * checkpoints.
 *
 * For example, given an input like:
 * ```graphql
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * will return the following SQL:
 * ```sql
 * DROP TABLE IF EXISTS checkpoint;
 * CREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));
 * INSERT checkpoint SET number = 0;Â·
 * DROP TABLE IF EXISTS votes;
 * CREATE TABLE votes (
 *   id VARCHAR(128) NOT NULL,
 *   name VARCHAR(128) NOT NULL,
 *   PRIMARY KEY (id) ,
 *   INDEX id (id),
 *   INDEX name (name)
 * );
 * ```
 *
 */
export function toSql(typeDefs: string): string {
  let sql = 'DROP TABLE IF EXISTS checkpoint;';
  sql += '\nCREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));';
  sql += '\nINSERT checkpoint SET number = 0;';

  getObjectsSchemas(typeDefs).forEach(type => {
    sql += `\n\nDROP TABLE IF EXISTS ${type.name.toLowerCase()}s;`;
    sql += `\nCREATE TABLE ${type.name.toLowerCase()}s (`;
    let sqlIndexes = ``;
    type.fields.forEach(field => {
      const fieldType = (field.type as IntrospectionNamedTypeRef).name;
      let sqlType = 'VARCHAR(128)';
      if (fieldType === 'Int') sqlType = 'INT(128)';
      if (fieldType === 'String') sqlType = 'VARCHAR(128)';
      if (fieldType === 'Text') sqlType = 'TEXT';
      sql += `\n  ${field.name} ${sqlType} NOT NULL,`;
      if (fieldType !== 'Text') sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
    });
    sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);`;
  });
  return sql;
}

const generateGqlQueries = (types: IntrospectionObjectType[]): string => {
  let gql = 'type Query {';
  types.forEach(type => {
    gql += `\n  ${type.name.toLowerCase()}s(`;
    gql += `\n    first: Int`;
    gql += `\n    skip: Int`;
    gql += `\n    orderBy: String`;
    gql += `\n    orderDirection: String`;
    gql += `\n    where: Where${type.name}`;
    gql += `\n  ): [${type.name}]`;
    gql += `\n  ${type.name.toLowerCase()}(id: String): ${type.name}`;
  });
  gql += `\n}`;
  return gql;
};

const generateGqlQueryInputs = (types: IntrospectionObjectType[]): string => {
  let where = '';
  types.forEach(type => {
    where += `\n\ninput Where${type.name} {`;
    type.fields.forEach(field => {
      const fieldType = (field.type as IntrospectionNamedTypeRef).name;
      if (fieldType !== 'Text') {
        where += `\n  ${field.name}: ${fieldType}`;
        where += `\n  ${field.name}_in: [${fieldType}]`;
        if (fieldType === 'Int') {
          where += `\n  ${field.name}_gt: ${fieldType}`;
          where += `\n  ${field.name}_gte: ${fieldType}`;
          where += `\n  ${field.name}_lt: ${fieldType}`;
          where += `\n  ${field.name}_lte: ${fieldType}`;
        }
      }
    });
    where += `\n}`;
  });
  return where;
};

/**
 * toGql returns a graphql schema string with generated queries for fetching
 * the already defined types in the input schema. For each type, a single and
 * multi (pluralized name) query is generated.
 *
 * For example, given the input schema:
 * ```
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * The generated queries will be like:
 * ```
 * type Query {
 *  votes(
 *     first: Int
 *     skip: Int
 *     orderBy: String
 *     orderDirection: String
 *     where: WhereVote
 *   ): [Vote]
 *   vote(id: String): Vote
 * }
 *
 *  input WhereVote {
 *    id: null
 *    id_in: [null]
 *    name: String
 *    name_in: [String]
 *  }
 *
 * ```
 *
 */
export function toGql(typeDefs: string): string {
  const objectTypes = getObjectsSchemas(typeDefs);
  const gqlQuery = generateGqlQueries(objectTypes);
  const gqlQueryInputs = generateGqlQueryInputs(objectTypes);

  return `${gqlQuery}\n${gqlQueryInputs}\n\n${typeDefs}`;
}

/**
 * toQuery creates an object of resolvers  based on types defined within graphql
 * schema.
 *
 * For example, given the input:
 * ```graphql
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * will return an object of the form:
 * ```
 * Object {
 *  vote: GQLResolver,
 *  votes: GQLResolver,
 * }
 * ```
 *
 * Note that a plural resolver is also created for each field
 */
export function toQueryResolver(typeDefs: string): IResolvers {
  const query = {};
  getObjectsSchemas(typeDefs).forEach(type => {
    query[`${type.name.toLowerCase()}s`] = queryMulti;
    query[type.name.toLowerCase()] = querySingle;
  });
  return query;
}
