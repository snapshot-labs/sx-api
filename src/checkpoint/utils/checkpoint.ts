import { CheckpointConfig } from '../types';

export const getContractFromCheckpointConfig = (config: CheckpointConfig): string[] => {
  return config.sources.map(source => source.contract);
};
