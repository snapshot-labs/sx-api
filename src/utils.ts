import fetch from 'node-fetch';
import { CallData, Contract, Provider, hash, shortString } from 'starknet';
import { Contract as EthContract } from '@ethersproject/contracts';
import { JsonRpcProvider } from '@ethersproject/providers';
import { faker } from '@faker-js/faker';
import { getAddress } from '@ethersproject/address';
import { BigNumber } from '@ethersproject/bignumber';
import { utils } from '@snapshot-labs/sx';
import {
  Space,
  StrategiesParsedMetadataItem,
  VotingPowerValidationStrategiesParsedMetadataItem
} from '../.checkpoint/models';
import EncodersAbi from './abis/encoders.json';
import ExecutionStrategyAbi from './abis/executionStrategy.json';
import SimpleQuorumExecutionStrategyAbi from './abis/l1/SimpleQuorumExecutionStrategy.json';
import { networkNodeUrl, networkProperties } from './overrrides';
import { handleStrategiesParsedMetadata } from './ipfs';

const ethProvider = new JsonRpcProvider(
  process.env.L1_NETWORK_NODE_URL ?? 'https://rpc.brovider.xyz/5'
);
const starkProvider = new Provider({
  rpc: {
    nodeUrl: networkNodeUrl
  }
});

const encodersAbi = new CallData(EncodersAbi);

export function getCurrentTimestamp() {
  return Math.floor(Date.now() / 1000);
}

export function toAddress(bn) {
  try {
    return getAddress(BigNumber.from(bn).toHexString());
  } catch (e) {
    return bn;
  }
}

export function getUrl(uri, gateway = 'pineapple.fyi') {
  const ipfsGateway = `https://${gateway}`;
  if (!uri) return null;
  if (
    !uri.startsWith('ipfs://') &&
    !uri.startsWith('ipns://') &&
    !uri.startsWith('https://') &&
    !uri.startsWith('http://')
  )
    return `${ipfsGateway}/ipfs/${uri}`;
  const uriScheme = uri.split('://')[0];
  if (uriScheme === 'ipfs') return uri.replace('ipfs://', `${ipfsGateway}/ipfs/`);
  if (uriScheme === 'ipns') return uri.replace('ipns://', `${ipfsGateway}/ipns/`);
  return uri;
}

export async function getJSON(uri) {
  const url = getUrl(uri);
  return fetch(url).then(res => res.json());
}

export function getSpaceName(address) {
  const seed = parseInt(hash.getSelectorFromName(address).toString().slice(0, 12));
  faker.seed(seed);
  const noun = faker.word.noun(6);
  return `${noun.charAt(0).toUpperCase()}${noun.slice(1)} DAO`;
}

export function dropIpfs(metadataUri: string) {
  return metadataUri.replace('ipfs://', '');
}

export function longStringToText(array: string[]): string {
  return array.reduce((acc, slice) => acc + shortString.decodeShortString(slice), '');
}

export function findVariant(value: { variant: Record<string, any> }) {
  const result = Object.entries(value.variant).find(([, v]) => typeof v !== 'undefined');
  if (!result) throw new Error('Invalid variant');

  return {
    key: result[0],
    value: result[1]
  };
}

export function getVoteValue(label: string) {
  if (label === 'Against') return 0;
  if (label === 'For') return 1;
  if (label === 'Abstain') return 2;

  throw new Error('Invalid vote label');
}

export async function handleExecutionStrategy(address: string, payload: string[]) {
  try {
    if (address === '0x0') return null;

    const executionContract = new Contract(ExecutionStrategyAbi, address, starkProvider);

    const executionStrategyType = shortString.decodeShortString(
      await executionContract.get_strategy_type()
    );

    let quorum = 0n;
    if (executionStrategyType === 'SimpleQuorumVanilla') {
      quorum = await executionContract.quorum();
    } else if (executionStrategyType === 'EthRelayer') {
      const [l1Destination] = payload;

      const SimpleQuorumExecutionStrategyContract = new EthContract(
        l1Destination,
        SimpleQuorumExecutionStrategyAbi,
        ethProvider
      );

      quorum = (await SimpleQuorumExecutionStrategyContract.quorum()).toBigInt();
    }

    return {
      executionStrategyType,
      quorum
    };
  } catch (e) {
    console.log('failed to get execution strategy type', e);

    return null;
  }
}

export async function updateProposaValidationStrategy(
  space: Space,
  validationStrategyAddress: string,
  validationStrategyParams: string[],
  metadataUri: string[]
) {
  space.validation_strategy = validationStrategyAddress;
  space.validation_strategy_params = validationStrategyParams.join(',');
  space.voting_power_validation_strategy_strategies = [];
  space.voting_power_validation_strategy_strategies_params = [];
  space.voting_power_validation_strategy_metadata = longStringToText(metadataUri);

  if (
    utils.encoding.hexPadLeft(validationStrategyAddress) ===
    utils.encoding.hexPadLeft(networkProperties.propositionPowerValidationStrategyAddress)
  ) {
    const parsed = encodersAbi.parse(
      'proposition_power_params',
      validationStrategyParams
    ) as Record<string, any>;

    if (Object.keys(parsed).length !== 0) {
      space.proposal_threshold = parsed.proposal_threshold;
      space.voting_power_validation_strategy_strategies = parsed.allowed_strategies.map(
        strategy => `0x${strategy.address.toString(16)}`
      );
      space.voting_power_validation_strategy_strategies_params = parsed.allowed_strategies.map(
        strategy => strategy.params.map(param => `0x${param.toString(16)}`).join(',')
      );
    }

    try {
      await handleVotingPowerValidationMetadata(
        space.id,
        space.voting_power_validation_strategy_metadata
      );
    } catch (e) {
      console.log('failed to handle voting power strategies metadata', e);
    }
  }
}

export async function handleStrategiesMetadata(
  spaceId: string,
  metadataUris: string[],
  startingIndex: number,
  type:
    | typeof StrategiesParsedMetadataItem
    | typeof VotingPowerValidationStrategiesParsedMetadataItem = StrategiesParsedMetadataItem
) {
  for (let i = 0; i < metadataUris.length; i++) {
    const metadataUri = metadataUris[i];

    const index = startingIndex + i;
    const uniqueId = `${spaceId}/${index}/${dropIpfs(metadataUri)}`;

    const exists = await type.loadEntity(uniqueId);
    if (exists) continue;

    const strategiesParsedMetadataItem = new type(uniqueId);
    strategiesParsedMetadataItem.space = spaceId;
    strategiesParsedMetadataItem.index = index;

    if (metadataUri.startsWith('ipfs://')) {
      strategiesParsedMetadataItem.data = dropIpfs(metadataUri);

      await handleStrategiesParsedMetadata(metadataUri);
    }

    await strategiesParsedMetadataItem.save();
  }
}

export async function handleVotingPowerValidationMetadata(spaceId: string, metadataUri: string) {
  if (!metadataUri) return;

  const metadata: any = await getJSON(metadataUri);
  if (!metadata.strategies_metadata) return;

  await handleStrategiesMetadata(
    spaceId,
    metadata.strategies_metadata,
    0,
    VotingPowerValidationStrategiesParsedMetadataItem
  );
}
