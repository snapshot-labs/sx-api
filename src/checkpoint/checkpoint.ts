import { GetBlockResponse, Provider } from 'starknet';
import { starknetKeccak } from 'starknet/utils/hash';
import { validateAndParseAddress } from 'starknet/utils/address';
import Promise from 'bluebird';
import getGraphQL from './graphql';
import { GqlEntityController } from './graphql/controller';
import { createLogger, Logger, LogLevel } from './utils/logger';
import { AsyncMySqlPool, createMySqlPool } from './mysql';
import { CheckpointConfig, CheckpointOptions } from './types';
import { getContractFromCheckpointConfig } from './utils/checkpoint';

export default class Checkpoint {
  public config;
  public writer;
  public schema: string;
  public provider: Provider;

  private readonly entityController: GqlEntityController;
  private readonly log: Logger;
  private readonly sourceContracts: string[];

  private mysqlPool?: AsyncMySqlPool;
  private mysqlConnection?: string;

  constructor(config: CheckpointConfig, writer, schema: string, opts?: CheckpointOptions) {
    this.config = config;
    this.writer = writer;
    this.schema = schema;
    this.entityController = new GqlEntityController(schema);
    this.provider = new Provider({ network: this.config.network });
    this.sourceContracts = getContractFromCheckpointConfig(config);

    this.log = createLogger({
      base: { component: 'checkpoint' },
      level: opts?.logLevel || LogLevel.Error,
      ...(opts?.prettifyLogs
        ? {
            transport: {
              target: 'pino-pretty'
            }
          }
        : {})
    });

    this.mysqlConnection = opts?.dbConnection;
  }

  /**
   * Returns an express handler that exposes a GraphQL API to query entities defined
   * in the schema.
   *
   */
  public get graphql() {
    return getGraphQL(this.entityController.createEntityQuerySchema(), {
      log: this.log.child({ component: 'resolver' }),
      mysql: this.mysql
    });
  }

  public async start() {
    this.log.debug('starting');
    const blockNum = await this.getStartBlockNum();
    return await this.next(blockNum);
  }

  public async reset() {
    this.log.debug('reset');
    await this.entityController.createEntityStores(this.mysql);
  }

  private async getStartBlockNum() {
    let start = 0;
    const lastBlock = await this.mysql.queryAsync('SELECT * FROM checkpoint LIMIT 1');
    const nextBlock = lastBlock[0].number + 1;
    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  private async getNextBlockNumber(blockNumber: number) {}

  private async next(blockNum: number) {
    //this.mysql.queryAsync(
    //'SELECT * FROM _checkpoints WHERE block_number = ? AND contract_address IN ?',
    //[blockNum, [this.sourceContracts]]
    //);
    this.log.debug({ blockNumber: blockNum }, 'next block');

    let block: GetBlockResponse;
    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');

      await Promise.delay(12e3);
      return this.next(blockNum);
    }

    await this.handleBlock(block);
    const query = 'UPDATE checkpoint SET number = ?';
    await this.mysql.queryAsync(query, [block.block_number]);
    return this.next(blockNum + 1);
  }

  private async handleBlock(block: GetBlockResponse) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    for (const receipt of Object.values(block.transaction_receipts)) {
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    }

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  private async handleTx(block, tx, receipt) {
    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction');

    for (const source of this.config.sources) {
      const contract = validateAndParseAddress(source.contract);

      if (contract === validateAndParseAddress(tx.contract_address)) {
        if (tx.type === 'DEPLOY' && source.deploy_fn) {
          this.log.info(
            { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
            'found deployment transaction'
          );

          await this.writer[source.deploy_fn]({ source, block, tx, receipt, mysql: this.mysql });
        }
      }

      for (const event of receipt.events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              await this.writer[sourceEvent.fn]({ source, block, tx, receipt, mysql: this.mysql });
            }
          }
        }
      }
    }

    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction done');
  }

  private get mysql(): AsyncMySqlPool {
    if (this.mysqlPool) {
      return this.mysqlPool;
    }

    // lazy initialization of mysql connection
    return (this.mysqlPool = createMySqlPool(this.mysqlConnection));
  }
}
