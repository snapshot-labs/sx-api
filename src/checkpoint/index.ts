import { Provider } from 'starknet';
import { starknetKeccak } from 'starknet/utils/hash';
import { validateAndParseAddress } from 'starknet/utils/address';
import Promise from 'bluebird';
import mysql from './mysql';
import getGraphQL from './graphql';
import { GqlEntityController } from './graphql/controller';
import { createLogger, Logger, LogLevel } from './utils/logger';

export interface CheckpointOptions {
  // Set the log output levels for checkpoint. Defaults to Silent.
  // Note, this does not affect the log outputs in writers.
  logLevel?: LogLevel;
  // optionally convert logs to pretty output.
  // will require installing pino-pretty. Not recommended for production.
  prettifyLogs?: boolean;
}

export default class Checkpoint {
  public config;
  public writer;
  public schema: string;
  public graphql;
  public provider: Provider;
  public checkpoints: number[];

  private readonly entityController: GqlEntityController;
  private readonly log: Logger;

  constructor(config, writer, schema: string, checkpoints: number[], opts?: CheckpointOptions) {
    this.config = config;
    this.writer = writer;
    this.schema = schema;
    this.entityController = new GqlEntityController(schema);
    this.graphql = getGraphQL(this.entityController.createEntityQuerySchema());
    this.provider = new Provider({ network: this.config.network });
    this.checkpoints = checkpoints;

    this.log = createLogger({
      base: { component: 'checkpoint' },
      level: opts?.logLevel || LogLevel.Silent,
      ...(opts?.prettifyLogs
        ? {
            transport: {
              target: 'pino-pretty'
            }
          }
        : {})
    });
  }

  async getStartBlockNum() {
    let start = 0;
    const lastBlock = await mysql.queryAsync('SELECT * FROM checkpoint LIMIT 1');
    const nextBlock = lastBlock[0].number + 1;
    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  async start() {
    this.log.debug('starting');
    const blockNum = await this.getStartBlockNum();
    return await this.next(blockNum);
  }

  async next(blockNum: number) {
    const cps = this.checkpoints.filter(cp => cp >= blockNum);
    if (cps.length > 0) blockNum = cps[0];
    let block: any;

    this.log.debug({ blockNumber: blockNum }, 'next block');

    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');

      await Promise.delay(12e3);
      return this.next(blockNum);
    }
    await this.handleBlock(block);
    const query = 'UPDATE checkpoint SET number = ?';
    await mysql.queryAsync(query, [block.block_number]);
    return this.next(blockNum + 1);
  }

  async handleBlock(block) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    for (const receipt of block.transaction_receipts) {
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    }

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  async handleTx(block, tx, receipt) {
    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction');

    for (const source of this.config.sources) {
      const contract = validateAndParseAddress(source.contract);

      if (contract === validateAndParseAddress(tx.contract_address)) {
        if (tx.type === 'DEPLOY' && source.deploy_fn) {
          this.log.info(
            { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
            'found deployment transaction'
          );

          await this.writer[source.deploy_fn]({ source, block, tx, receipt });
        }
      }

      for (const event of receipt.events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              await this.writer[sourceEvent.fn]({ source, block, tx, receipt });
            }
          }
        }
      }
    }

    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction done');
  }

  async reset() {
    this.log.debug('reset');
    await this.entityController.createEntityStores(mysql);
  }
}
