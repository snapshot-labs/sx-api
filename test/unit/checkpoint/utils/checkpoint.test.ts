import { getContractFromCheckpointConfig } from '../../../../src/checkpoint/utils/checkpoint';
import { validCheckpointConfig } from '../../../fixtures/checkpointConfig.fixture';

describe('getContractFromCheckpointConfig', () => {
  it('should work', () => {
    expect(getContractFromCheckpointConfig(validCheckpointConfig)).toMatchSnapshot();
  });
});
