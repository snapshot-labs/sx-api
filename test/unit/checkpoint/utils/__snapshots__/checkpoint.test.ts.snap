// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContractFromCheckpointConfig should work 1`] = `
Array [
  "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
]
`;
